src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -153,39 +154,61 @@ object JoinIndexRule
+      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
         val updatedOutput =
           baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
         baseRelation.copy(relation = relation, output = updatedOutput)
     }
 
-    val complementIndexPlan = plan transformUp {
+    val hybridScanEnabled = spark.sessionState.conf
+      .getConfString(
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED,
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED_DEFAULT.toString)
+      .toBoolean
+
+    if (hybridScanEnabled) {
+      getComplementIndexPlan(index, plan, replacedPlan)
+    } else {
+      replacedPlan
+    }
+  }
+
+  private def getComplementIndexPlan(
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    val complementIndexPlan = originalPlan transformUp {
       case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
         val updatedOutput =
-          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
 
         val files =
-          fsRelation.location.inputFiles.map(new Path(_))
+          fsRelation.location.inputFiles
+            .map(new Path(_))
             .filter(!index.allSourceFileSet.contains(_))
-        val newLocation =
-          new InMemoryFileIndex(spark, files, Map(), None)
-        val newRelation =
-          fsRelation.copy(location = newLocation, dataSchema = index.schema)(spark)
 
-        val newLogicalRelation = baseRelation.copy(relation = newRelation, output = updatedOutput)
-
-        val attrs = newLogicalRelation.attributeMap.baseMap
-          .filter { key =>
-            index.indexedColumns contains key._2._2.name
-          }
-          .map { case (_, ref) => ref._1 }
-        // TODO fix
-        //  cur: project => filter => exchange => relation
-        //  desired: exchange => project => filter => relation
-        RepartitionByExpression(attrs.toSeq, newLogicalRelation, index.numBuckets)
+        if (files.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, files, Map(), None)
+          val newRelation =
+            fsRelation.copy(location = newLocation, dataSchema = index.schema)(spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
     }
 
-    BucketUnionLogicalPlan(Seq(replacedPlan, complementIndexPlan), bucketSpec)
+    if (!originalPlan.equals(complementIndexPlan)) {
+      val attrs = complementIndexPlan.output.attrs.filter { attr =>
+        index.indexedColumns contains attr.name
+      }
+      val shuffled = RepartitionByExpression(attrs, complementIndexPlan, index.numBuckets)
+      BucketUnionLogicalPlan(
+        Seq(indexPlan, shuffled),
+        index.bucketSpec.copy(sortColumnNames = Seq()))
```

Comment:
could you please add a comment here to show why ordering is removed? something like ` // removing sort order because original data isn't sorted by the same columns as the index`. Thanks